File: Python/Longest Valid Parentheses.py
class Solution:
    # @param s, a string
    # @return an integer
    def longestValidParentheses(self, s):
        dp = [0] * (len(s)+1)
        for i in range(len(s)):    
            if s[i] == "(":
                continue
            else:  
                j = i-dp[i]-1   
                if j >= 0 and s[j] == "(":
                    dp[i+1] = dp[i] + 2
                    if j > 0:
                        dp[i+1] += dp[j]
            print dp
        return max(dp)
s = Solution()
print s.longestValidParentheses("()(())(())")

File: Java/src/Longest Valid Parentheses.java
import java.util.ArrayList;
import java.util.List;

public class Solution
{
  public int longestValidParentheses(String s)
  {
    List<Integer> stack = new ArrayList<Integer>();
    int max = 0;
    int left = -1;
    for (int i = 0; i < s.length(); i++)
    {
      if (s.charAt(i) == '(')
        stack.add(i);
      else
      {
        if (stack.isEmpty())
          left = i;
        else
        {
          stack.remove(stack.size() - 1);
          if (stack.isEmpty())
            max = Math.max(max, i - left);
          else
            max = Math.max(max, i - stack.get(stack.size() - 1));
        }
      }
    }
    return max;
  }
}
